return_type.md

```diff
@@ -0,0 +1,94 @@
+    return_type(f, T::Type{<:Tuple}) -> R::Type
+
+Return a type `R` such that `f(args...) isa R` where `args isa T`.
+
+The type `R` obtained from `return_type` is merely an upper bound.  There may
+exist a stricter type `S` such that `f(args...) isa S` for every `args isa T`
+with `S <: R` and `S != R`. Furthermore, the exact type `R` obtained from
+`return_type` depends on various factors including but not limited to the exact
+Julia version used, packages loaded, and command line options. As such, when
+used in publicly registered packages, **it is the package authors'
+responsibility to ensure that the API guarantees provided by the package do not
+depend on the exact type `R` obtained from `return_type`.**
+
+# Extended help
+
+## Examples
+
+The following function is an invalid use-case of `return_type`.
+
+```julia
+"""
+    invalid_usecase1(f, xs::AbstractArray) -> ys::Array
+
+Return an array `ys` such that `vec(ys)` is `isequal`-equivalent to
+
+    [f(xs[1]), f(xs[2]), ..., f(xs[end])]
+"""
+function invalid_usecase1(f, xs)
+    R = return_type(f, Tuple{eltype(xs)})
+    ys = similar(xs, R)
+    for i in eachindex(xs, ys)
+        ys[i] = f(xs[i])
+    end
+    return ys
+end
+```
+
+This is because the value obtained through `eltype(invalid_usecase1(f, xs))`
+depends on exactly what `return_type` returns.  It may be fixed by re-computing
+the element type before returning the result.
+
+```julia
+function valid_usecase1(f, xs)
+    R = return_type(f, Tuple{eltype(xs)})
+    ys = similar(xs, R)
+    S = Union{}
+    for i in eachindex(xs, ys)
+        ys[i] = f(xs[i])
+        S = promote_type(S, typeof(ys[i]))
+    end
+    if S != R
+        zs = similar(xs, S)
+        copyto!(zs, ys)
+        return zs
+    end
+    return ys
+end
```

Comment:
I'm not sure if this function is a good example usecase of `return_type`.
A reasonable/robust API guarantee in Julia will likely end up requiring runtime information (as we compute `S` here), while `return_type` should basically be used  merely for optimization purposes (as the compiler does type inference merely for optimization purposes). 

So I think it might be more reasonable to incorporate your `isconcretetype` example here  to explain how we _can_ use `return_type` for some optimization, e.g.:
```julia
function valid_usecase1(f, xs)
    R = return_type(f, Tuple{eltype(xs)})
    if Base.isconcretetype(R)
        # optimizable case
        ys = similar(xs, R)
        for i in eachindex(xs, ys)
            ys[i] = f(xs[i])
        end
        return ys
    else
        # fallback case
        ys = similar(xs, R)
        S = Union{}
        for i in eachindex(xs, ys)
            ys[i] = f(xs[i])
            S = promote_type(S, typeof(ys[i]))
        end
        if S != R
            zs = similar(xs, S)
            copyto!(zs, ys)
            return zs
        end
        return ys
    end
end
```

---

And probably one obvious exception of this view ("`return_type` for optimization") would be the usecase for computing element type of an empty container. We would need to mention that API guarantee should be loosened in such situations as we naively mention in the `promote_op` docstring?

Reply:
Hmm... I rather hope that the (future) compiler can elide runtime `S = promote_type(S, typeof(ys[i]))` for `isconcretetype(R)` with maybe `ys[i]` manually hoisted out. I think this is possible at least in FLoops/Transducers that try to unroll the loop until the accumulator type-stabilizes. But yes, I agree your example is a more down-to-earth example for the current compiler and it's better to use it here.